src/main/java/io/divolte/server/filesinks/FileFlusher.java

```diff
@@ -0,0 +1,295 @@
+     * Optional.empty() and healthy otherwise. The file system is marked unhealthy
+     * on any IOException that is thrown by the used file manager for this sink.
+     * File system recovery is attempted on heart beats of the processor. Processing
+     * is paused as soon as the file system goes into unhealthy state. As a result,
+     * process(...) can assume the Optional to be populated; all other methods need
+     * to check for file system health by inspecting the state of the optional
+     * before performing any operation. The lastFixAttempt field is used to keep
+     * track of the time of the recent most reconnect attempt in order to implement
+     * a back off larger than the heart beat frequency.
+     */
+    private Optional<TrackedFile> currentTrackedFile;
+    private long lastFixAttempt;
+
+    public FileFlusher(final FileStrategyConfiguration configuration, final FileManager manager) {
+        this(configuration, manager, DEFAULT_FILE_SYSTEM_RECONNECT_DELAY);
+    }
+
+    public FileFlusher(final FileStrategyConfiguration configuration, final FileManager manager, final long reconnectDelay) {
+        /*
+         * Constructor with configurable reconnect delay for testability.
+         */
+        this.reconnectDelay = reconnectDelay;
+
+        syncEveryMillis = configuration.syncFileAfterDuration.toMillis();
+        syncEveryRecords = configuration.syncFileAfterRecords;
+        newFileEveryMillis = configuration.rollEvery.toMillis();
+
+        instanceNumber = INSTANCE_COUNTER.incrementAndGet();
+        hostString = findLocalHostName();
+
+        this.manager = Objects.requireNonNull(manager);
+
+        try {
+            currentTrackedFile = Optional.of(new TrackedFile(manager.createFile(newFileName())));
+        } catch(final IOException ioe) {
+            // Postpone throwing the exception to force going into heartbeat / recover
+            // cycle. Potentially drops a record too many, but avoids a additional branch in
+            // process(...).
+            currentTrackedFile = Optional.of(brokenTrackedFile(ioe));
+        }
+    }
+
+    @Override
+    public ProcessingDirective process(final Item<AvroRecordBuffer> item) {
+        final long time = System.currentTimeMillis();
+        try {
+            final TrackedFile trackedFile = currentTrackedFile.get();
+            trackedFile.divolteFile.append(item.payload);
+            trackedFile.recordsSinceLastSync += 1;
+
+            possiblySyncAndOrRoll(time);
+
+            return CONTINUE;
+        } catch(final IOException ioe) {
+            markFileSystemUnavailable(time);
+            logger.error("File system connection error. Marking file system as unavailable. Attempting reconnect after " + reconnectDelay + " ms.", ioe);
+
+            return PAUSE;
+        }
+    }
+
+    @Override
+    public ProcessingDirective heartbeat() {
+        final long timeMillis = System.currentTimeMillis();
```

Comment:
OK. Changed everything to use nanosecond granularity and `System.nanoTime()` to keep track of elapsed times. Small down side is that the timestamps are not really interesting logging information anymore, but they only appear in debug logging so far.